[1단계 - 자동차 경주 구현] 말랑(신동훈) 미션 제출합니다.

shin-mallang opened the pull request:
안녕하세요 카일 ~!
`말랑` & `채채` 입니다.

코드와 설계에 많은 부족한 부분들이 있을텐데, 잘 부탁드립니다!

미션을 진행하면서 추가로 궁금했던 점은 코멘트로 남겨두겠습니다!

chaewon121 committed:
test: String 클래스에 대한 학습 테스트

chaewon121 committed:
test: Set 클래스에 대한 학습 테스트

chaewon121 committed:
docs: 기능 명세서 작성

chaewon121 committed:
feat: 자동차 이름 구현

chaewon121 committed:
feat: 자동차의 위치 구현

chaewon121 committed:
feat: 랜덤숫자 생성 기능 구현

chaewon121 committed:
feat: 이름에 대한 equals & hashcode 구현

chaewon121 committed:
feat: 위치에 대한 equals & hashcode 구현

chaewon121 committed:
test: 깨진 테스트 수정

chaewon121 committed:
feat: 자동차 도메인 구현

chaewon121 committed:
refactor: add final keyword

chaewon121 committed:
feat: 자동차들을 관리하는 Cars 구현

chaewon121 committed:
feat: 바퀴수를 관리하는 Lap 구현

chaewon121 committed:
test: String의 split() 메서드 학습 테스트 추가

chaewon121 committed:
feat: 기본 게임 단계 인터페이스 & 추상클래스 생성

chaewon121 committed:
feat: 자동차 이름을 입력받는 View 구현

chaewon121 committed:
feat: 문자열을 받아 자동차를 생성하는 컨트롤러 기능 구현

chaewon121 committed:
feat: 자동차를 입력받는 단계 구현

chaewon121 committed:
feat: 전체 바퀴수 입력받는 View 구현

chaewon121 committed:
feat: 전체 바퀴수를 생성하는 Controller 기능 구현

chaewon121 committed:
feat: 전체 바퀴수를 입력받는 단계 구현

chaewon121 committed:
feat: "실행 결과"를 출력하는 View 구현

chaewon121 committed:
feat: "실행 결과"를 출력하는 단계 구현

chaewon121 committed:
refactor: 컨트롤러 테스트 이름 상세한 예시 작성

chaewon121 committed:
feat: 자동차들을 움직이는 컨트롤러 기능 구현

chaewon121 committed:
refactor: 하드코딩한 문자열대신 상수로 사용

chaewon121 committed:
feat: 현재 상태를 출력하는 View 구현

chaewon121 committed:
feat: 자동차를 움직이는 단계 구현

chaewon121 committed:
feat: 자동차들 중 우승한 자동자를 반환하는 기능 구현

chaewon121 committed:
feat: 우승자 이름리스트를 반환하는 컨트롤러 기능 구현

chaewon121 committed:
feat: 우승자를 보여주는 단계 구현

chaewon121 committed:
feat: 게임 종료 단계 구현

chaewon121 committed:
feat: 게임 메인 메서드 구현 (기능 구현 완료)

shin-mallang committed:
feat: 예외처리 기능 추가

shin-mallang committed:
refactor: Step에서 Cars와 Lap을 넘기는 대신 Context를 공유하도록 수정

shin-mallang committed:
refactor: NumberGenerator를 인터페이스로 분리

shin-mallang committed:
refactor: 테스트 클래스에 @DisplayName 추가

shin-mallang committed:
refactor: Config를 통해 설정하도록 구현

shin-mallang committed:
refactor: Context의 생성을 static method를 통해 생성하여 의도를 드러내도록 변경

shin-mallang committed:
style: import 정리, 라인 다듬기

shin-mallang committed:
test: @Test 가 생략된 테스트에 어노테이션 추가

shin-mallang committed:
refactor: 예외 횟수가 3회를 초과하면 예외를 발생하는 대신 정상 종료하도록 변경

shin-mallang committed:
refactor: 우승자 출력 시 이름 순서대로 출력하도록 수정

shin-mallang commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -57,16 +58,33 @@ void test_3() {
+        List<Car> winners = cars.winners();
+
+        // then
+        assertThat(winners).containsExactlyInAnyOrder(winner1, winner2);
```
`Cars`의 생성자로는 `List<String>`만을 받기 때문에, 우승자를 가려내는 기능을 위해 자동차들의 이동 상태를 임의로 지정하는 방법이 필요하였습니다.

`Cars`의 생성자에서 `List<Car>`을 받도록 변경하여 이를 편하게 테스트 하게 만들 수도 있었지만, `Car`들의 생성 역시 `Cars`의 책임 중 하나라 생각하여 이렇게 변경하지 않고 `Reflection API`를 사용하였습니다.

혹시 테스트 코드 상에서 `Reflection`을 통해 private 필드 값에 접근하여 임의로 값을 바꾸는 것에 대해 어떻게 생각하시나요?

KIMSIYOUNG replied:
필요에 따라 사용할 수 있을 것 같아요. 다만 위 경우에는 우리는 이미 자동차를 움직이는 로직중 랜덤한 값을 사용하지 않게끔 잘 구현하셨는데, 직접 이동시키지 않고 리플랙션을 사용해야하는 이유가 있을까요?

shin-mallang committed:
refactor: Name의 필드를 name에서 value로 수정

KIMSIYOUNG commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,24 @@
+        return position;
+    }
+
+    public void move(final int randomNumber) {
```
자동차 입장에서 전달받는 값이 랜덤이라는 것을 알 필요가 있을까요?

shin-mallang replied:
임의의 값으로 움직인다는 것에 집중하다 보니, 임의의 값임을 알도록 하는 것이 좋겠다 생각했었는데, 말씀을 듣고 보니 임의의 값이라는 것은 Car에서는 몰라도 되는 값인 것 같네요! 감사합니다!

KIMSIYOUNG commented on src/main/java/racingcar/domain/Position.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    private Position nextPosition(final int randomNumber) {
+        if (randomNumber >= BOUNDARY_NUMBER) {
```
Position 이 이동할지 말지를 결정하는게 자연스럽나요?

shin-mallang replied:
Position을 위치를 나타내는 값이자, 위치와 관련된 책임을 갖는 하나의 객체로 생각했었던 것 같아요! 말씀해주신 것을 듣고 보니 Postion은 그 자체로 값을 표현하는 것으로만 사용하는 것이 좋을 것 같다는 생각이 드네요!
한번 수정해 보도록 하겠습니다😊

KIMSIYOUNG commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,49 @@
+        return this.cars;
+    }
+
+    public List<Car> winners() {
```
Winner를 판단하는 것도 객체로 포장해주면 좋을 것 같아요.

shin-mallang replied:
오 좋은 방법인 것 같아요 👍
지금 떠오르는 방법은 전략 패턴을 통해 승자를 판단하는 기능을 구현하면 좋을 것 같다고 생각이 드는데 어떻게 생각하시나요?

하지만 전체적인 리뷰를 들었을 때 제 코드가 과도하게 복잡하게 구현되어 있는 부분이 많아 간단하게 구현해 보라는 말씀이 있어서 해당 부분도 간단하게 구현해서 수정해 보도록 할게요!

KIMSIYOUNG replied:
승자를 판단하는 로직이 전략패턴으로 구현되었을 때 장점이 있을까요? 현재 승자를 판단하는 로직은 하나뿐이고, NumberGenerator와 같이 테스트할 수 없는 랜덤한 값이 있는 것도 아닌데 전략패턴으로 구현될 이유는 없을 것 같아요.

너무 어려운 주제지만 어디까지 유연하게 만들것인가에서 저는 현재의 요구사항을 만족하는 최소한의 코드가 가장 잘 작성된 코드라고 생각되어요. 말랑은 변화에 유연한 코드가 잘 작성된 코드라고 생각한다면 전략패턴으로 구현하셔도 괜찮습니다. 저는 개인적인 경험상 제가 예상하는 것과 어플리케이션이 발전, 수정되는 방향이 일치하지 않는 경우가 많아서 성급한 추상화는 오히려 가독을 해치는 경험을 많이해서 위와 같은 의견을 가지고 있는 것 같아요.

shin-mallang replied:
아하 저는 항상 코드를 구현할 때, 모든 프로그램은 변화할 수 있기 때문에 변경에 대비할 수 있도록 코드를 작성하자는 생각이 조금 강하게 있는 것 같아요! 
그런데 이번 카일의 리뷰를 통해 어떤 상황에서도 변경에 대처 가능한 코드를 작성하기보다는, 현재 해결해야하는 문제 상황에 맞는 구조를 설계하는 것이 중요하다고 느꼈어요!
변경에 유연한 코드가 우선 구현하는데도 더 복잡하고, 가독성 측면에서도 별로 좋지 못하다는 단점을 가지고 있기 때문에 이 모든 것은 트래이드 오프라는 생각이 들었네요!
게다가 첫 코드리뷰이다보니 제가 지금까지 공부한 것들을 녹여내고 싶어서 조금 더 복잡하게 짰던 부분도 있는 것 같아요!
좋은 말씀 너무 감사해요 카일 😊

KIMSIYOUNG commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,49 @@
+                .collect(Collectors.toList());
+    }
+
+    private List<Car> sortedCars() {
```
네이밍을 보고 내림차순 정렬이라는 것을 알 수 있나요? 추가로 정렬이 필요한건가요, 최대값이 필요한건가요?

shin-mallang replied:
네이밍은 수정하는게 좋을 것 같습니다! 좋은 지적 감사합니다!

혹시 다음 질문의 의도를 잘 파악하지 못하였는데, 조금 더 자세히 말씀해주실 수 있으실까요?
> 추가로 정렬이 필요한건가요, 최대값이 필요한건가요?

KIMSIYOUNG replied:
최대값을 찾아서, 우승자를 찾아내기 위한 로직인데 정렬을 하는 기능을 제공할 필요가 있는가? 라는 질문이었습니다!

KIMSIYOUNG reviewed (changes requested):
안녕하세요! 전체적으로 잘 구현해주셨네요. 
전체적으로 객체 분리, 일급컬렉션 사용, 불변 객체 등 이미 많은 것을 학습하고 오신 것 같아요.
궁금한 점과 제안드리고 싶은 내용이 있는데요. 

1.  현재 코드는 페어와 동의된 코드인가요? 혹은 한분이 리딩해서 작성한 코드인가요?
2. 현재의 프로그램이 이렇게 복잡하게 구현되어야 하는 프로그램인가요? 상태패턴과 스프링 형태의 코드를 제외하고 구현해보시고 현재의 코드와 비교해보시면 좋을 것 같습니다.   

위 내용을 고민해보시고, 같이 이야기 나눈 이후에 추가 리뷰를 진행하는게 좋을 것 같습니다.

shin-mallang commented:
좋은 리뷰 너무나 감사합니다😊!

우선 첫번째 질문에 대해서는 따로 DM으로 답변드리고 싶은데 가능할까요?

두번째 의견에 대해서는 다음과 같은 질문을 드리고 싶습니다!
- 언급해주신 `스프링 형태의 코드`에서 어느 부분을  보시고 `스프링 형태`라고 생각하셨는지 궁금합니다! 
`Config`를 보고 말씀해주신 것인가요? 또는 다른 부분이 있다면 조금 더 자세히 설명해 주시면 감사하겠습니다!

카일이 말씀해주신것을 보고 나니 조금 과도할 정도로 간단한 프로그램을 복잡하게 구현했다는 생각이 들어 반성하게 되네요.
조금 불필요하게 복잡한 부분들은 제거해서 다시 구현해 보도록 할게요!

KIMSIYOUNG commented:
네네 물론이죠 디엠으로 얘기나눠도 좋습니다!

- 네네 config 형태를 보고 말씀드렸습니다. 다른 부분은 크게 느끼지 않았던 것 같아요.

shin-mallang committed:
feat: Position 값을 증가시키는 기능 구현

shin-mallang committed:
refactor: 자동차를 이동시키를 책임을 Car에서 담당하도록 구현

shin-mallang committed:
refactor: 경주의 우승자를 구하는 테스트 코드의 중복 제거

shin-mallang committed:
refactor: 경주의 우승자를 구하는 기능을 담당하는 객체를 변경

shin-mallang committed:
refactor: 도메인 패키지를 car와 game으로 분리

shin-mallang committed:
refactor: 테스트를 위한 Mock과 Util 클래스 패키지 분리

shin-mallang committed:
feat: 게임과 관련된 기능을 담당하는 클래스 구현

shin-mallang committed:
refactor: Step을 제거하고, 게임의 진행을 Controller에서 담당하도록 구현

shin-mallang committed:
fix: 패키지 변경으로 인해 깨지는 테스트 수정

shin-mallang committed:
refactor: 로그 기능을 인터페이스를 사용하지 않고 구현

shin-mallang committed:
feat: 반복 관련 상태를 담당하는 Retry 구현

shin-mallang committed:
feat: 예외 처리 진행

shin-mallang committed:
refactor: 매직 넘버를 의미있는 상수로 변경

shin-mallang committed:
feat: 테스트를 위한 Reflection 편의 기능 구현

shin-mallang committed:
feat: 테스트를 위한 Reflection 편의 기능 구현

shin-mallang committed:
refactor: Lap을 불변 값 타입으로 설정

shin-mallang committed:
refactor: 예외처리 로직 수정

shin-mallang committed:
refactor: NumberGenerator 인터페이스 대신 클래스로 구현

shin-mallang committed:
refactor: Cars에서 이동 후 결과 상태를 반환하도록 구현

Cars의 내부 상태를 숨기기 위해(getter 를 public 으로 공개하지 않도록) Cars에서 반환하는 것이 더 좋겠다 판단하여 이렇게 수정하였습니다.
이에 대한 카일의 의견이 궁금해요!

shin-mallang commented:
카일!
카일이 말씀해주신대로 과도하게 복잡하게 구현하였던 부분을 덜어내고 간단하게 구현해 보았습니다!
코드를 이해하기도 훨씬 쉬워졌고, 구현하는 입장에서도 편했어요 😊

앞으로는 각각 구현해야하는 애플리케이션의 규모와 상황에 맞춰서 적절하게 구현하는 연습을 해보아야 할 것 같아요!

이번 리팩토링을 진행하면서 고민했던 부분이 하나 있는데, 해당 부분에 대한 카일의 의견을 듣고 싶어요 

리팩토링 과정 중에서  마지막 커밋인 `refactor: Cars에서 이동 후 결과 상태를 반환하도록 구현` 에 제 고민과, 선택에 대한 이유를 적어두었어요!

Cars.move()의 결과 데이터를 담은 `MovedResult` 클래스를 구현하였는데, `MovedResult`를 Cars에서 생성해야 할지, `RacingCarGame`에서 생성해야 할 지 고민이 되었습니다.

처음에는 `이동 결과를 보여주는 것`은 게임의 규칙이기 때문에, 게임과 관련된 책임을 수행하는 `RacingCarGame`에서 만들어 반환하는 것이 맞다고 생각을 해서 `RacingCarGame`에서 생성하도록 구현하였는데요,

이렇게 하니 Cars의 getter가 public으로 노출되어 외부에서 Cars 내부에 속한 `List<Car>`에 들어있는 `Car`들의 상태를 변경할 수 있게 되었어요.

저는 그래서 위치는 `RacingCarGame`에 있는 것이 더 좋다고 생각하지만, Cars의 내부 상태를 보호하여 사이드 이팩트를 방지하기 위해 Cars에서 반환하도록 하였습니다.

생각하는 다른 방법으로는 Cars의 getter를 public으로 공개하되, 이때 반환되는 리스트를 깊은 복사를 통해 반환하는 방법도 생각해 보았어요!

카일은 이중에 어느 방법이 좋다고 생각하는지, 카일의 생각과 이유가 궁금해요!
혹은 다른 좋은 방법이 있을까요? 🧐

아 그리고 첫번째 PR에 커멘트로 추가한 부분인데, private field의 값을 설정하거나 접근하기 위해 Reflection API를 사용하는 것에 대해 어떻게 생각하시나요?

shin-mallang commented:
헉 완전 사고를 쳤네요..ㅠㅠ😭
맨 처음 커멘트를 달 때, Pending 라벨을 못 보고 커멘트를 남겼는데, 방금 Pending 라벨이 붙은 것에 대해 의문을 갖고 찾아보니 큰일이 났다는 것을 알아버렸어요 ㅠㅠ
죄송해요 카일..ㅠㅠ 지금에서야 커멘트를 제대로 남기네요..😥

KIMSIYOUNG commented on docs/README.md:
```diff
@@ -0,0 +1,31 @@
+
+### 자동차
+
+- [x] 0에서 9 사이의 random 값에 따라 움직인다.
```
문서를 잘 작성해주셨네요👏

KIMSIYOUNG commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -0,0 +1,63 @@
+        this.winnerJudge = winnerJudge;
+    }
+
+    public void gameStart() {
```
흐름도가 깔끔하게 잘 보이네요.

RacingCarGame 이라는 객체가 꼭 필요한지도 고민해보면 좋을 것 같아요. 제거해도 충분히 이해하기 좋은 코드라고 생각이 들어요.

shin-mallang replied:
게임이라는 것 자체를 하나의 도메인이라 생각했습니다.
Cars에서는 단지 자동차들을 관리하게끔만 하고, 그러한 자동차들을 이용해서 경주시키고, 경주의 최종 결과를 반환하는 것은 게임의 책임이라고 판단했어요.

KIMSIYOUNG commented on src/main/java/racingcar/controller/RacingCarController.java:
```diff
@@ -0,0 +1,63 @@
+            MovedResult movedResult = game.race();
+            OutputView.printState(movedResult);
+        }
+        GameResult gameResult = game.gameResult();
```
GameResult 라는 네이밍 보단 Winner정도로 풀어낼 수도 있을 것 같아요.

많은 것을 함축할 수 있는 네이밍이 좋을 때도 있지만 대체로는 해당 의미를 명확하게 드러낼 수 있는 네이밍이 가독에는 훨씬 용이한 것 같아요.

shin-mallang replied:
GameResult보다는 카일의 말씀대로 Winner가 우승자라는 의미를 명확하게 드러낼 수 있어서 좋다고 생각해요!
감사합니다😊

KIMSIYOUNG commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -0,0 +1,42 @@
+package racingcar.domain.car;
+
+public class Car {
```
자동차 객체 깔끔하게 잘 구현해주셨네요. 

말씀주신대로 이동한다를 판단하고 실행하는 역할은 위치보단, 자동차에서 수행하는 것이 자연스러운 것 같아요.

KIMSIYOUNG commented on src/main/java/racingcar/domain/car/Cars.java:
```diff
@@ -0,0 +1,42 @@
+    private static final int MIN_SIZE = 2;
+    private final List<Car> cars;
+
+    public Cars(final List<String> cars) {
```
변수명은 carNames 인 것 같네요🙂

shin-mallang replied:
부끄럽네요.. ㅠㅠ😥

KIMSIYOUNG commented on src/main/java/racingcar/domain/car/Cars.java:
```diff
@@ -0,0 +1,42 @@
+        return this.cars;
+    }
+
+    public List<Car> sortedCarsByPositionDesc() {
```
이 기능을 Cars 에서 제공해줘야하는건지 고민해보면 좋을 것 같아요. 

승자를 구하는 로직을 위해 해당 기능을 열어주어야 하는지, 승자를 구하는 곳에서 해당 기능을 구현해야하는지 생각해보시면 좋을 것 같습니다.

KIMSIYOUNG commented on src/main/java/racingcar/domain/car/Cars.java:
```diff
@@ -0,0 +1,42 @@
+
+public class Cars {
+
+    private static final int MIN_NUMBER = 0;
```
RacingGame이라는 것이 없다면 이 곳이 적절한 위치겠지만 있다면 해당 상수는 여기에 있는게 적절하지 않은 것 같아요.

shin-mallang replied:
저도 그렇게 생각해요!
코드를 리팩토링하는 과정에서 미쳐 확인하지 못한 것 같아요!
좋은 지적 감사합니다!

KIMSIYOUNG commented on src/main/java/racingcar/domain/car/MovedResult.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class MovedResult {
```
해당 객체를 기존에는 DTO로 간주하셨던 것 같은데 이번에는 도메인에 들어온 이유가 있나요?

shin-mallang replied:
> 현재 자동차와 MovedResult 의 의존성이 양방향으로 잡혀있는 것 같습니다...

위의 커멘트에 대한 답변과 같은 답변을 드릴 수 있을 것 같고, 동일하게 다음 PR에서 구조를 개선해 보도록 할게요!

KIMSIYOUNG commented on src/main/java/racingcar/domain/game/Lap.java:
```diff
@@ -0,0 +1,47 @@
+
+import java.util.Objects;
+
+public class Lap {
```
객체들을 잘 포장해주셨네요👏

shin-mallang replied:
감사합니다! 칭찬 들으니 너무 기쁘네요😊

KIMSIYOUNG commented on src/main/java/racingcar/domain/game/NumberGenerator.java:
```diff
@@ -0,0 +1,10 @@
+package racingcar.domain.game;
+
+public class NumberGenerator {
```
요 부분은 난수를 테스트할 수 없기에 전략패턴으로 분리되어도 괜찮은 것 같아요🙂

KIMSIYOUNG commented on src/main/java/racingcar/domain/game/WinnerJudge.java:
```diff
@@ -0,0 +1,21 @@
+
+    private static final int WINNER_INDEX = 0;
+
+    public List<Car> judge(final Cars cars) {
```
객체는 자신이 가지고 있는 데이터를 기반으로 어떠한 역할을 수행할 수 있는 주체라고 생각하고 있어요.

현재는 객체라기보단 static한 형태와 다를게 없는 것 같아요.

Winners라는 일급컬렉션도 자동차들을 가지고 있고 그들중 위너를 판단하는 로직을 가질 수 있을 것 같고 그것이 조금 더 자연스러워 보이네요.

shin-mallang replied:
Winners를 만들 생각을 하지 못했는데, 정말 좋은 생각인 것 같아요!
`sortedCarsByPositionDesc()` 메서드에 대한 커멘트에 대한 답변으로도, 이를 Cars에서 제공할 것이 아니라 Winners에서 내부적으로 구현하면 될 것 같다고 생각해요!

KIMSIYOUNG commented on src/test/java/racingcar/mock/MockFixedNumberGenerator.java:
```diff
@@ -0,0 +1,17 @@
+
+import racingcar.domain.game.NumberGenerator;
+
+public class MockFixedNumberGenerator extends NumberGenerator {
```
상속보단 인터페이스나 조합을 통한 구현이 더 좋을 것 같아요. 조금 더 학습해보시면 좋을 것 같습니다.

KIMSIYOUNG commented on src/main/java/racingcar/domain/game/RacingCarGame.java:
```diff
@@ -0,0 +1,59 @@
+        return new RacingCarGame(numberGenerator, winnerJudge, cars, lap);
+    }
+
+    /**
```
`Getter` 를 사용하지 않는 이유가 어떤 것이라고 생각하시나요? 정말 그 자체를 절대규칙으로 지켜야하는 것인지, 무엇을 위해 저런 말이 있는건지 고민해보면 좋을 것 같습니다.

shin-mallang replied:
객체가 자신의 상태를 스스로 책임지도록 하기 위해 getter를 사용하지 않는 것을 권장한다고 생각해요!
getter를 쓰면 모든 곳에서 해당 객체의 상태를 마음대로 변경할 수 있게끔 되니까요.

저 혼자 코드를 작성하는 경우라면 getter를 제공하더라도, 객체가 자신의 상태를 스스로 책임지게끔만 구현하면 딱히 문제가 되지 않는다고 생각해요.

다만 저는 발생하지 않을 상황에 대해서 너무 걱정이 많았던 것 같아요!
getter를 public으로 선언하고, 다른 개발자와 함께 협업하는 상황에서는 다른 개발자가 getter를 사용하여 객체가 자신의 상태를 책임지지 못하게 만들어 버릴 수도 있다고 생각했어요!

이번에 카일의 리뷰를 받으며, 너무 여러 상황에 대해서 과도하게 걱정하는 것을 조금 고쳐야 할 필요가 있다고 생각했어요!

지금까지 혼자 공부하며 해결하지 못하고 있던 것들이 카일의 리뷰를 통해 어느정도 해소된 것 같아요!

정말 감사합니다! 

다음 PR에서는 해당 부분에 대해서도 반영해 보도록 할게요! 😊

KIMSIYOUNG commented on src/main/java/racingcar/domain/car/Cars.java:
```diff
@@ -0,0 +1,42 @@
+                .collect(Collectors.toUnmodifiableList());
+    }
+
+    public MovedResult move(final NumberGenerator generator) {
```
현재 자동차와 MovedResult 의 의존성이 양방향으로 잡혀있는 것 같습니다.
단방향으로 정리하되, 어떤 변화가 생겼을 때 어디를 고쳐야하는가를 기준으로 고민해보시면 좋을 것 같습니다.

예를들어 도메인 로직에 변경이 생겼을 때 누굴 고쳐야하는가, 화면에 이런 값도 노출해야할 때 어디를 고쳐야하는가 등

shin-mallang replied:
이 부분에 대해서는 혼자 코드를 작성하며 많은 고민을 했었고, 해결을 못하고 있었었어요!

MovedResult는 아래의 comment에서 카일이 말씀해 주신 것과 같이 DTO로 여전히 간주하고 있는 것이 맞습니다.

그러나 이를 Controller 패키지에 둔다면 domain에서 controller에 의존하게 된다고 생각했어요!

그래서 어쩔 수 없이 이를 MovedResult에서 구현하려 했습니다.

사실 이 모든 것이 getter를 사용하지 않고 코드를 작성하려 했던 이유에서였는데요, 아래에서 카일이 말씀해주신 getter를 지향해야하는 이유에 대해 생각해 보니, 이 부분을 어떻게 개선해야 할 지에 대한 고민도 자연스레 해결된 것 같아요!

다음 PR에서 수정하도록 하겠습니다.

정말 감사드려요 카일🥹

KIMSIYOUNG reviewed (approved):
안녕하세요 말랑!
훨씬 가독성 좋은 코드가 된 것 같아요.
1단계에서 학습할 내용은 이미 잘 인지하고 계신 것 같습니다.
몇가지 피드백 남겼는데 확인해보시고 2단계에서 추가 진행하면 좋을 것 같아요!

shin-mallang reviewed (commented):
카일 정말 너무나도 감사했던 리뷰였습니다.
저에게 엄청나게 많은 도움이 된 것 같아요!
지금까지 혼자 공부하면서 고민했던 많은 부분들에 대해서 카일의 리뷰를 통해 다시금 생각하게 되었고, 어느정도 해결된 것 같아요!
정말 너무 감사했고, 최고였습니다 카일! 😊
